crates/holochain/src/test_utils/consistency.rs

```diff
@@ -0,0 +1,479 @@
+        })
+        .await;
+}
+async fn check_expected_data(
+    reporter: Reporter,
+    session: ConsistencySession,
+    vault: EnvRead,
+    p2p_env: EnvRead,
+) {
+    if let Err(e) = check_expected_data_inner(reporter.clone(), session, vault, p2p_env).await {
+        reporter
+            .send_report(SessionReport::Error {
+                error: e.to_string(),
+            })
+            .await;
+    }
+}
+
+fn generate_session<'iter>(
+    all_agents: &'iter Vec<(Arc<KitsuneAgent>, DhtArc)>,
+    all_hashes: &'iter Vec<Arc<KitsuneOpHash>>,
+    timeout: Duration,
+    agent_env_map: HashMap<Arc<KitsuneAgent>, EnvRead>,
+) -> impl Iterator<Item = (Arc<KitsuneAgent>, ConsistencySession, EnvRead)> + 'iter {
+    all_agents
+        .iter()
+        .map(move |(agent, arc)| {
+            let mut published_hashes = Vec::new();
+            let mut expected_agents = Vec::new();
+            for hash in all_hashes.iter() {
+                if arc.contains(hash.get_loc()) {
+                    published_hashes.push(hash.clone());
+                }
+            }
+            for (agent, _) in all_agents.iter() {
+                if arc.contains(kitsune_p2p::KitsuneBinType::get_loc(&(**agent))) {
+                    expected_agents.push(agent.clone());
+                }
+            }
+            (
+                agent.clone(),
+                ExpectedData {
+                    expected_agents,
+                    expected_hashes: published_hashes,
+                },
+            )
+        })
+        .map(move |(agent, expected_data)| {
+            (
+                agent,
+                ConsistencySession {
+                    keep_alive: Some(Duration::from_secs(1)),
+                    frequency: Duration::from_millis(100),
+                    timeout,
+                    expected_data,
+                },
+            )
+        })
+        .filter_map(move |(agent, expected_session)| {
+            agent_env_map
+                .get(&agent)
+                .cloned()
+                .map(|env| (agent, expected_session, env))
+        })
+}
+async fn check_expected_data_inner(
+    reporter: Reporter,
+    session: ConsistencySession,
+    vault: EnvRead,
+    p2p_env: EnvRead,
+) -> DatabaseResult<()> {
+    let ConsistencySession {
+        keep_alive,
+        frequency,
+        timeout,
+        expected_data:
+            ExpectedData {
+                expected_agents,
+                mut expected_hashes,
+            },
+    } = session;
+
+    let start = tokio::time::Instant::now();
+    let mut last_keep_alive = start;
+    let deadline = tokio::time::Instant::now() + timeout;
+    let mut frequency = tokio::time::interval(frequency);
+    let mut missing_agents = Vec::with_capacity(expected_agents.len());
+    let mut missing_hashes = Vec::with_capacity(expected_hashes.len());
+    while let Ok(_) = tokio::time::timeout_at(deadline, frequency.tick()).await {
```

Comment:
Wouldn't this `deadline` timeout be applied independently to every tick, not to the overall while loop?

Reply:
A deadline is a point in time that when we've past any future will timeout immediately. It's not relative to the time the future was called.